core/src/main/java/com/graphhopper/storage/index/LocationIndexTree.java

```diff
@@ -564,6 +572,133 @@ protected boolean check(int node, double normedDist, int wayIndex, EdgeIteratorS
+     *        wide/high (with center tile containing (queryLat, queryLon)). Note that it is
+     *        possible that the search radius exceeds this square - in this case, an exception
+     *        will be thrown.
+     */
+    public List<QueryResult> findWithinRadius(final double queryLat, final double queryLon,
+            final EdgeFilter edgeFilter, final double radius, final Integer maxIterations) {
+
+        // check the max iterations:
+        int requiredIterationsForThisRadius = (int) Math.ceil(radius / this.minResolutionInMeter)
+                + 1; // note that this is a conservative upper bound
+        int _maxIterations;
+        if (maxIterations == null) {
+            // used the maximum iterations required:
+            _maxIterations = requiredIterationsForThisRadius;
+        } else {
+            // check the provided maxIterations will include this radius:
+            if (maxIterations < requiredIterationsForThisRadius)
+                throw new IllegalArgumentException("A minimum of " + requiredIterationsForThisRadius
+                        + " iterations are required for a radius of " + radius + "m, but a "
+                        + "maxIterations of " + maxIterations + " was provided. Behaviour is "
+                        + "undefined in this case, so please either decrease maxIterations to "
+                         + requiredIterationsForThisRadius + ", or decrease the radius.");
+            _maxIterations = maxIterations;
+        }
+        
+        final double returnAllResultsWithin = distCalc.calcNormalizedDist(radius);
+        // implement a cheap priority queue via List, sublist and Collections.sort
+        final List<QueryResult> queryResults = new ArrayList<QueryResult>();
+        GHIntHashSet set = new GHIntHashSet();
+
+        for (int iteration = 0; iteration < _maxIterations; iteration++) {
+            // TODO: should we use the return value of earlyFinish?
+            findNetworkEntries(queryLat, queryLon, set, iteration);
+            final GHBitSet exploredNodes = new GHTBitSet(new GHIntHashSet(set));
+            final EdgeExplorer explorer = graph.createEdgeExplorer(edgeFilter);
+            set.forEach(new IntPredicate() {
+                @Override
+                public boolean apply(int node) {
+                    new XFirstSearchCheck(queryLat, queryLon, exploredNodes, edgeFilter) {
+                        @Override
+                        protected double getQueryDistance() {
+                            // do not skip search if distance is 0 or near zero (equalNormedDelta)
+                            return Double.MAX_VALUE;
+                        }
+
+                        @Override
+                        protected boolean check(int node, double normedDist, int wayIndex,
+                                EdgeIteratorState edge, QueryResult.Position pos) {
+                            if (normedDist < returnAllResultsWithin || queryResults.isEmpty()
```

Comment:
The idea was to have at least one candidate even if outside, so maybe this is not necessary, yes. 

Here again let us try to minimize this issue and create improvement issues separately to make merging this faster.